### PR TITLE
Feature/201 automatic setup qa engine in google cloud

### LINKED ISSUE #201: Automatic Setup QA engine in Google Cloud
- [x] #258 

- [ ] Automate set up QA engine

### PR DESCRIPTION
**Pull request by Jorge-Dacal**
Closes #201

### CODE REVIEW

**Review by albertdb — CHANGES_REQUESTED** [review]
Good overall. I will send suggestions ASAP. Blocking meanwhile.

File: scripts/sonarqube/gcloud/set-terraform-variables.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+file="./terraform.tfvars"
+../../set-config.sh --file_set_vars $file $*
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/gcloud/set-terraform-variables.sh:3]: This scripts call another second script and Sonatype is asking us to enclose with quotation marks the argument that will pass from first script to the second all the arguments with which it has been invoked. If we add this quotation which would make only one input argument, generating a bug in the second script.

[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/gcloud/set-terraform-variables.sh:3]: Solved all but not the same four that are in the first comment, that are repeated here.

These four scripts call another second script and Sonatype is asking us to enclose with quotation marks the argument that will pass from first script to the second all the arguments with which it has been invoked. If we add this quotation which would make only one input argument, generating a bug in the second script.

[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/gcloud/set-terraform-variables.sh:3]: @sonatype-lift ignoreall

[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/gcloud/set-terraform-variables.sh:3]: @sonatype-lift ignoreall

File: scripts/sonarqube/common/create_token.sh [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        -p | --password)       password=$2; shift 2;;
+        -t | --token-name)     tokenName=$2; shift 2;;
+        -h | --help)           helpFunction; exit ;;
+        --tf-output-sq-token)  terraformSQToken=$2; terraformGetSQToken; shift 2;;
+        --) shift; break;;
+    esac
+done
+
+white='\e[1;37m'
+green='\e[1;32m'
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/common/create_token.sh]: fixed

File: scripts/sonarqube/common/create_token.sh [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        sonar_ready=true
+    else
+        ((attempt+=1))
+        sleep 10
+    fi
+done
+
+# Generate token if sonarqube is ready
+if $sonar_ready; then
+    setToken=$(curl -sS -H "Content-Type: application/x-www-form-urlencoded" -d "name=$tokenName" -u $user:$password $sonarUrl/api/user_tokens/generate)
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/common/create_token.sh]: fixed

File: scripts/sonarqube/common/create_token.sh [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        --) shift; break;;
+    esac
+done
+
+white='\e[1;37m'
+green='\e[1;32m'
+
+# Check url provided
+if [[ -z "$sonarUrl" ]]; then
+    echo -e "${red}Error: Missing paramenters, -s or --sonar-url is mandatory." >&2
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/common/create_token.sh:53]: fixed

File: scripts/sonarqube/common/create_token.sh [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+# To use it add the flag --tf-output-sq-token $outputName when invoking the script.
+function terraformGetSQToken() {
+    FILE="terraform.tfstate"
+    if [[ -f "$FILE" ]]; then
+        if cp -f $FILE temp.tfstate; then
+           sq_token_output=$(terraform output -state=temp.tfstate | grep "$terraformSQToken" | cut -d' ' -f 3)
+           rm -f temp.tfstate
+        fi
+    fi
+    if [[ ! -z "$sq_token_output" ]]; then
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/common/create_token.sh]: fixed

File: scripts/sonarqube/common/create_token.sh [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+fi
+if [[ -z "$tokenName" ]]; then
+    tokenName="hangarToken"
+fi
+
+# Check sonar status
+sonar_ready=false
+attempt=0
+while ! $sonar_ready && [[ $attempt -lt 30 ]]; do
+    health=$(curl -sS --retry 30 --max-time 10 --retry-connrefused -u $user:$password $sonarUrl/api/system/health)
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/sonarqube/common/create_token.sh]: fixed

**Review by ultymatom — DISMISSED** [review]

**Review by lpinon — DISMISSED** [review]
Working fine! Only issue needs to be inside the sonar folder in order to work.

File: scripts/tf.sh [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    echo "Common flags:"
+    echo ""
+    echo "  -s, --state-folder    The folder where you are going to save/import your terraform configuration."
+    echo "  -k, --output-key      [ONLY FOR output] The key of the terraform output variable that you want to recover."
+    echo "  -q, --quiet           To not print any command of the script, only the execution of terraform \$command."
+    echo "  -h, --help            Get help for commands."
+}
+
+vars=""
+while [[ -n "${@}" ]]; do
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: changed the while condition to
while [[ $# -gt 0 ]]; do

**Review by lpinon — APPROVED** [review]

**Review by ultymatom — DISMISSED** [review]

File: scripts/tf.sh [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    echo "tf.sh [command] [flags] [terraform variables]"
+    echo ""
+    echo "Main commands:"
+    echo ""
+    echo "  COMMAND       DESCRIPTION"
+    echo "  apply         Create or update infrastructure"
+    echo "  destroy       Destroy previously-created infrastructure"
+    echo "  output        Show output values from your terraform. With output command only one flag is readed `--output-key` or `-k`, all other flags are ignored."
+    echo "                  If you want to show the output as json, add an option `--output-key --json`."
+    echo "                  If you want to recover only one output value add an option `--output-key \$key` where \$key is the name of the output var."
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: removed the $ to avoid this issue

[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: remove $ to avoid this issue

File: scripts/tf.sh [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    echo "Usage"
+    echo ""
+    echo "tf.sh [command] [flags] [terraform variables]"
+    echo ""
+    echo "Main commands:"
+    echo ""
+    echo "  COMMAND       DESCRIPTION"
+    echo "  apply         Create or update infrastructure"
+    echo "  destroy       Destroy previously-created infrastructure"
+    echo "  output        Show output values from your terraform. With output command only one flag is readed `--output-key` or `-k`, all other flags are ignored."
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: changed `` for '' to avoid this issue

[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: changed `` for '' to avoid this issue

**Review by lpinon — APPROVED** [review]

**Review by ultymatom — APPROVED** [review]

**Review by lpinon — APPROVED** [review]

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (albertdb) [inline .vscode/settings.json:1]: Remove this file

Comment (Jorge-Dacal) [reply .vscode/settings.json]: removed

File: documentation/src/common_templates/setup-sonarqube-instance.asciidoc [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+IMPORTANT: In windows, keep in mind that after applying any changes, you will lose the value of the token so be sure to copy or write it down before applying any changes. To avoid this we have implemented a method but to work you must store the standard terraform output in a file called terraform.tfoutput. This can be done with the following command:
+
+```
+terraform output > terraform.tfoutput
+```
+
+== Change Sonarqube default admin password
+
+After having deployed sonarqube by following this guide, you will be able to access SonarQube web interface on the url provided by terraform output and the following credentials:
```
Comment (albertdb) [inline documentation/src/common_templates/setup-sonarqube-instance.asciidoc:150]:
```suggestion
After a few minutes, you will be able to access SonarQube web interface on the public URL provided by Terraform output with the following credentials:
```

File: documentation/src/common_templates/setup-sonarqube-instance.asciidoc [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+* Username:   `admin`
+* Password:   `admin`
+
+IMPORTANT: Change the default password promptly. After that, update the password in terraform vars, you can do it manually or with the next command:
+
+```
+./set-terraform-variables.sh --sonarqube_password ${YOUR_NEW_PASSWORD}
+```
```
Comment (albertdb) [inline documentation/src/common_templates/setup-sonarqube-instance.asciidoc:159]:
```suggestion
IMPORTANT: Change the default password promptly. After that, update the password in Terraform configuration: `./set-terraform-variables.sh --sonarqube_password <new password>`.
```

File: documentation/src/common_templates/setup-sonarqube-instance.asciidoc [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+== Appendix: More information about terraform for {provider_name}
```
Comment (albertdb) [inline documentation/src/common_templates/setup-sonarqube-instance.asciidoc:161]:
```suggestion
== Appendix: More information about Terraform for {provider_name}
```

File: documentation/src/gcloud/setup-sonarqube-instance.asciidoc [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+--subnet_cidr_block                  The range of internal addresses that are owned by this subnetwork. Ranges must be unique and non-overlapping within a network. By default: 10.0.1.0/29 + \
+-- instance_type                     Machine Instance type. By dafeault: e2-medium
+
+include::../common_templates/setup-sonarqube-instance.asciidoc[tag=partBeforePrerequisites]
+=== Prerequisites
+* https://developer.hashicorp.com/terraform/tutorials/gcp-get-started/install-cli[Install Terraform].
+
+* https://cloud.google.com/sdk/docs/install-sdk[Install Google Cloud CLI].
+
+* Enable Compute Engine in your gcloud project. You can do it via webconsole or via glcoud cli using the following command:
```
Comment (albertdb) [inline documentation/src/gcloud/setup-sonarqube-instance.asciidoc:21]:
```suggestion
* Enable Compute Engine in your Google Cloud project. You can do it via Web Console or with the following command:
```

File: documentation/src/gcloud/setup-sonarqube-instance.asciidoc [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+```
+gcloud services enable compute.googleapis.com --project "$projectName"
+```
+
+* Have a service account that is going to be used by terraform with the next roles:
```
Comment (albertdb) [inline documentation/src/gcloud/setup-sonarqube-instance.asciidoc:27]:
```suggestion
* Have the `key.json` file of a service account with (at least) the following roles:
```

File: documentation/src/gcloud/setup-sonarqube-instance.asciidoc [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+```
+roles/compute.instanceAdmin.v1
+roles/compute.networkAdmin
+roles/compute.securityAdmin
+```
+
+IMPORTANT: You need the key.json file of that service account to run terraform as it.
```
Comment (albertdb) [inline documentation/src/gcloud/setup-sonarqube-instance.asciidoc:35]:
```suggestion
```

File: documentation/src/gcloud/setup-sonarqube-instance.asciidoc [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+:terraform_tutorials: https://developer.hashicorp.com/terraform/tutorials/gcp-get-started
+:extra_information: https://cloud.google.com/docs/terraform[Official Gcloud documentation]
+:terraform_vars_example_short: --service_account_file key.json --project hangar
+:terraform_vars_example_full: --service_account_file key.json --project hangar --region europe-southwest1 --zone europe-southwest1-a --subnet_cidr_block 10.0.1.0/24 -- instance_type e2-medium
+:terraform_vars: --service_account_file    [Required] Json file with the google credentials. + \
+--project                 [Required] Short name (ID) of the project. + \
+--region                             The GCP region where the resources will be created. By default: europe-southwest1 + \
+--zone                               The zone inside the region that the machine should be created in. By default: europe-southwest1-a + \
+--subnet_cidr_block                  The range of internal addresses that are owned by this subnetwork. Ranges must be unique and non-overlapping within a network. By default: 10.0.1.0/29 + \
+-- instance_type                     Machine Instance type. By dafeault: e2-medium
```
Comment (albertdb) [inline documentation/src/gcloud/setup-sonarqube-instance.asciidoc:13]:
```suggestion
:terraform_vars: --service_account_file    [Required] JSON file of a service account with required roles. + \
--project                 [Required] Short name (ID) of the project. + \
--region                             Region where the resources will be created. Default: europe-southwest1 + \
--zone                               Zone inside the region where the resources will be created. Default: europe-southwest1-a + \
--subnet_cidr_block                  Range of internal addresses that are owned by this subnetwork. Ranges must be unique and non-overlapping within a network. Default: 10.0.1.0/29 + \
-- instance_type                     Machine Instance type. Default: e2-medium
```

File: documentation/src/common_templates/setup-sonarqube-instance.asciidoc [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+* {terraform_tutorials}[Official Terraform tutorials]
+ifdef::extra_information[]
+{extra_information}
```
Comment (albertdb) [inline documentation/src/common_templates/setup-sonarqube-instance.asciidoc:164]:
```suggestion
* {extra_information}
```

File: documentation/src/aws/setup-sonarqube-instance.asciidoc [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+:terraform_vars_example_short: --region eu-west-1 --keypair_name sonarqube
+:terraform_vars_example_full: --region eu-west-1 --vpc_cidr_block 10.0.0.0/16 --subnet_cidr_block 10.0.1.0/24 --nic_private_ip 10.0.1.50 --instance_type t3a.small --keypair_name sonarqube
+:terraform_vars: + \
+--region               The AWS region where the resources will be created. By default: eu-west-1 + \
+--vpc_cidr_block       Virtual private network IP range (CIDR). By default: 10.0.0.0/16 + \
+--subnet_cidr_block    The range of internal addresses that are owned by this subnetwork. Ranges must be unique and non-overlapping within a network. + \
+                       By default: 10.0.1.0/24 + \
+--nic_private_ip       Instance private IP within subnet range.  By default: 10.0.1.50 + \
+--instance_type        Machine Instance type. By default: t3a.small + \
+--keypair_name         Keypair name to connect with ssh as defined in AWS. By default: sonarqube
```
Comment (albertdb) [inline documentation/src/aws/setup-sonarqube-instance.asciidoc:15]:
```suggestion
:terraform_vars: --region               Region where the resources will be created. Default: eu-west-1 + \
--vpc_cidr_block       Virtual private network IP range (CIDR). Default: 10.0.0.0/16 + \
--subnet_cidr_block    Range of internal addresses that are owned by this subnetwork. Ranges must be unique and non-overlapping within a network. Default: 10.0.1.0/24 + \
--nic_private_ip       Instance private IP within subnet range.  Default: 10.0.1.50 + \
--instance_type        Machine Instance type. Default: t3a.small + \
--keypair_name         Keypair name to connect with SSH as defined in AWS. Default: sonarqube
```

File: scripts/tf.sh [hunk 20]
```diff
@@ -0,0 +1,149 @@
+#! /bin/bash
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/tf.sh]: renamed

File: scripts/accounts/gcloud/setup-principal-account.sh [hunk 21]
```diff
@@ -130,7 +130,7 @@ then
         fi
         echo -e "${white}Creating service-account keys for service account $service_account..."
         key_path="${key_path:-./key.json}"
-        if ! gcloud iam service-accounts keys create "${key_path}" --iam-account="$service_account" &> /dev/null;
+        if ! gcloud iam service-accounts keys create "${key_path}" --iam-account="$service_account_email" &> /dev/null;
```
[the remark this answers is not in the document]
Comment (Jorge-Dacal) [reply scripts/accounts/gcloud/setup-principal-account.sh:133]: it's ok

File: scripts/accounts/azure-devops/create-project.sh [hunk 22]
```diff
@@ -34,7 +34,7 @@ red='\e[0;31m'
 if [ -z "$projectName" ] || [ -z "$description" ] || [ -z "$organization" ] || [ -z "$visibility" ] || [ -z "$pat_token" ] 
 then
    echo -e "${red}Error: Some required parameters are missing." >&2
-   echo -e ${white} >&2
+   echo -e "${white}" >&2
```
Comment (Jorge-Dacal) [inline scripts/accounts/azure-devops/create-project.sh:37]:
```suggestion
   echo -ne "${white}" >&2
```

**Review by albertdb — APPROVED** [review]

File: .gitignore [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+.vscode/*
+
+# Terraform files
+**/.terraform/**
+**/.terraform.lock.hcl
+**/terraform.tfstate
+**/terraform.tfstate.backup
+
+# secrets folder
+**/.secrets/**
```
Comment (abcarcol) [inline .gitignore:11]: Is this file part of the PR?

**Review by abcarcol — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 23 above)
Comment (albertdb) [reply .gitignore:11]: I don't see why not.

**Review by lpinon — APPROVED** [review]

**Review by abcarcol — APPROVED** [review]

**Pull request merged**